woocommerce/woocommerce-ios
# [IAP] Error screens for upgrades flow

## Description

This PR deals with the different error state views and their navigation flow (if needed). There's 5 potential errors we're dealing with:

- Error when fetching plans
- Error when loading merchant entitlements
- Error when attempting to upgrade more than one site 
- Purchase error
- In-App Purchases not supported

## Testing instructions
The easiest way to check all error screens is to force the specific case as `default` within the `UpgradesView`, rather than switching through all cases, and leave the `.loading` case as well just to see the state switch when we retry switching between states:

```
switch upgradesViewModel.upgradeViewState {
  case .loading:
    OwnerUpgradesView(upgradePlan: .skeletonPlan(), purchasePlanAction: {}, isLoading: true)
  default:
    let upgradeError = UpgradesError.purchaseError
    UpgradesErrorView(upgradeError,
      onRetryButtonTapped: {},
      onRetryPaymentButtonTapped: {},
      onCancelUpgradeTapped: {})
.padding(Layout.padding)
}
```

Then just switch through the different errors and load the preview or run the app for testing the button callback:

1. The cases for `.fetchError`, and `.entitlementsError` will show a `Retry` button. Tapping the button will attempt to fetch plans again:
```
let upgradeError = UpgradesError.fetchError
let upgradeError = UpgradesError.entitlementsError
```

2.  The case `.purchaseError` will show a `Try Payment Again` button which will retry fetching plans again and offer the option to tap on "Purchase" if a merchant wants to upgrade (same behaviour as the errors above), and a `Cancel Upgrade` button, which dismisses the view and redirects the merchant back to My Store:
```
let upgradeError = UpgradesError.purchaseError
```
3. Will show the screens above, without any button to interact with.
```
let upgradeError = UpgradesError.inAppPurchasesNotSupported
let upgradeError = UpgradesError.maximumSitesUpgraded
```

## Screenshots
| `fetchError`, `entitlementsError` | `maximumSitesUpgraded`  | `purchaseError` | `inAppPurchasesNotSupported`|
|--------|--------|--------|--------|
| <img width=400 src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/7969e325-3ade-407a-8421-a5a2087960a9"> | <img width=400 src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/e876216f-2e3d-4c0f-ae87-c5d749d3cc3f"> | <img width=400 src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/622fb470-afff-4bde-a386-b0e3a883ef6c"> | <img width=400 src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/ee2d14dc-04f9-49fd-8887-766b6c7c764f"> |
